File: app/src/main/java/ru/p3tr0vich/fuel/SyncLocal.java
package ru.p3tr0vich.fuel;

import android.support.annotation.NonNull;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

class SyncLocal {

    private final SyncFiles mSyncFiles;

    SyncLocal(@NonNull SyncFiles syncFiles) {
        mSyncFiles = syncFiles;
    }

    private void load(@NonNull File file, @NonNull List<String> strings) throws IOException {
        UtilsFileIO.checkExists(file);
        UtilsFileIO.read(file, strings);
    }

    private void save(@NonNull File file, @NonNull List<String> strings) throws IOException {
        UtilsFileIO.createFile(file);
        UtilsFileIO.write(file, strings);
    }

    private int getRevision(@NonNull File fileRevision) {
        List<String> strings = new ArrayList<>();
        try {
            load(fileRevision, strings);

            return Integer.decode(strings.get(0));
        } catch (Exception e) {
            return -1;
        }
    }

    public int getPreferencesRevision() {
        return getRevision(mSyncFiles.getLocalFilePreferencesRevision());
    }

    public int getDatabaseRevision() {
        return getRevision(mSyncFiles.getLocalFileDatabaseRevision());
    }

    public void loadPreferences(@NonNull List<String> preferences) throws IOException {
        load(mSyncFiles.getLocalFilePreferences(), preferences);
    }

    public void loadDatabase(@NonNull List<String> syncRecords) throws IOException {
        load(mSyncFiles.getLocalFileDatabase(), syncRecords);
    }

    public void savePreferences(@NonNull List<String> preferences) throws IOException {
        save(mSyncFiles.getLocalFilePreferences(), preferences);
    }

    public void saveDatabase(@NonNull List<String> syncRecords) throws IOException {
        save(mSyncFiles.getLocalFileDatabase(), syncRecords);
    }

    public void saveDatabaseFullSync() throws IOException {
        UtilsFileIO.createFile(mSyncFiles.getLocalFileDatabaseFullSync());
    }

    private void saveRevision(@NonNull File fileRevision, int revision) throws IOException {
        List<String> strings = new ArrayList<>();
        strings.add(String.valueOf(revision));

        save(fileRevision, strings);
    }

    public void savePreferencesRevision(int revision) throws IOException {
        saveRevision(mSyncFiles.getLocalFilePreferencesRevision(), revision);
    }

    public void saveDatabaseRevision(int revision) throws IOException {
        saveRevision(mSyncFiles.getLocalFileDatabaseRevision(), revision);
    }

    public void makeDirs() throws IOException {
        UtilsFileIO.makeDir(mSyncFiles.getLocalDirDatabase());
        UtilsFileIO.makeDir(mSyncFiles.getLocalDirPreferences());
    }

    public void deleteFiles() throws IOException {
        UtilsFileIO.deleteFile(mSyncFiles.getLocalFilePreferences());
        UtilsFileIO.deleteFile(mSyncFiles.getLocalFilePreferencesRevision());

        UtilsFileIO.deleteFile(mSyncFiles.getLocalFileDatabase());
        UtilsFileIO.deleteFile(mSyncFiles.getLocalFileDatabaseFullSync());
        UtilsFileIO.deleteFile(mSyncFiles.getLocalFileDatabaseRevision());
    }

    public void deleteLocalFileDatabase() throws IOException {
        UtilsFileIO.deleteFile(mSyncFiles.getLocalFileDatabase());
    }
}

File: app/src/main/kotlin/ru/p3tr0vich/fuel/sync/SyncLocal.kt
package ru.p3tr0vich.fuel.sync

import ru.p3tr0vich.fuel.utils.UtilsFileIO
import ru.p3tr0vich.fuel.utils.UtilsLog
import java.io.File
import java.io.IOException
import java.util.*

internal class SyncLocal(private val syncFiles: SyncFiles) {

    val preferencesRevision: Int
        get() = getRevision(syncFiles.localFilePreferencesRevision)

    val databaseRevision: Int
        get() = getRevision(syncFiles.localFileDatabaseRevision)

    @Throws(IOException::class)
    private fun load(file: File, strings: MutableList<String>) {
        UtilsFileIO.checkExists(file)
        UtilsFileIO.read(file, strings)
    }

    @Throws(IOException::class)
    private fun save(file: File, strings: List<String>) {
        UtilsFileIO.createFile(file)
        UtilsFileIO.write(file, strings)
    }

    private fun getRevision(fileRevision: File): Int {
        val strings = ArrayList<String>()

        return try {
            load(fileRevision, strings)

            Integer.decode(strings[0])
        } catch (e: Exception) {
            UtilsLog.d(TAG, "getRevision", "addAccountExplicitly == false")
            -1
        }
    }

    @Throws(IOException::class)
    fun loadPreferences(preferences: MutableList<String>) {
        load(syncFiles.localFilePreferences, preferences)
    }

    @Throws(IOException::class)
    fun loadDatabase(syncRecords: MutableList<String>) {
        load(syncFiles.localFileDatabase, syncRecords)
    }

    @Throws(IOException::class)
    fun savePreferences(preferences: List<String>) {
        save(syncFiles.localFilePreferences, preferences)
    }

    @Throws(IOException::class)
    fun saveDatabase(syncRecords: List<String>) {
        save(syncFiles.localFileDatabase, syncRecords)
    }

    @Throws(IOException::class)
    private fun saveRevision(fileRevision: File, revision: Int) {
        val strings = ArrayList<String>()

        strings.add(revision.toString())

        save(fileRevision, strings)
    }

    @Throws(IOException::class)
    fun savePreferencesRevision(revision: Int) {
        saveRevision(syncFiles.localFilePreferencesRevision, revision)
    }

    @Throws(IOException::class)
    fun saveDatabaseRevision(revision: Int) {
        saveRevision(syncFiles.localFileDatabaseRevision, revision)
    }

    @Throws(IOException::class)
    fun makeDirs() {
        UtilsFileIO.makeDir(syncFiles.localDirDatabase)
        UtilsFileIO.makeDir(syncFiles.localDirPreferences)
    }

    @Throws(IOException::class)
    fun deleteFiles() {
        UtilsFileIO.deleteFile(syncFiles.localFilePreferences)
        UtilsFileIO.deleteFile(syncFiles.localFilePreferencesRevision)

        UtilsFileIO.deleteFile(syncFiles.localFileDatabase)
        UtilsFileIO.deleteFile(syncFiles.localFileDatabaseRevision)
    }

    companion object {
        private const val TAG = "SyncLocal"
    }
}